src/placeos-models/api_key.cr

```diff
@@ -0,0 +1,106 @@
+require "random"
+require "crypto/bcrypt"
+
+require "./base/model"
+require "./user_jwt"
+
+module PlaceOS::Model
+  class ApiKey < ModelBase
+    include RethinkORM::Timestamps
+
+    table :api_key
+
+    attribute name : String, es_subfield: "keyword"
+    attribute description : String = ""
+    attribute scopes : Array(String) = ["public"]
+
+    # when nil it defaults to the users permissions
+    attribute permissions : UserJWT::Permissions? = nil
+
+    attribute secret : String = ->{ Random::Secure.urlsafe_base64(32) }, mass_assignment: false
+
+    belongs_to User
+    belongs_to Authority
+
+    def user=(user)
+      super(user)
+      self.authority_id = user.authority_id
+    end
+
+    secondary_index :authority_id
+
+    # Validation
+    ###############################################################################################
+
+    # Ensure name is unique under the authority scope
+    #
+    ensure_unique :name, scope: [:authority_id, :name] do |authority_id, name|
+      {authority_id, name.strip.downcase}
+    end
+
+    # Callbacks
+    ###############################################################################################
+
+    before_create :safe_id
+    before_create :set_authority
+    before_create :hash!
+
+    # Reject '+' and '~'
+    protected def safe_id
+      self._new_flag = true
+      @id ||= Random.new.hex(16)
```

Comment:
I think this creates a hex string with 16 bytes to entropy, not a string of 16 chars length.

Reply:
nice, good catch

Reply:
ahh yeah so this should be 32 bytes long which seems good